junit5/src/main/java/cz/xtf/junit5/listeners/ProjectCreator.java

```diff
@@ -53,17 +106,71 @@ public void visit(NamespaceBuilder builder) {
         if (OpenShiftConfig.pullSecret() != null) {
             openShift.setupPullSecret(OpenShiftConfig.pullSecret());
         }
+        log.info("Created namespace: " + openShift.getNamespace());
+    }
+
+    @Override
+    public void afterAll(ExtensionContext context) {
+        if (OpenShiftConfig.useNamespacePerTestcase() && JUnitConfig.cleanOpenShift()) {
+            deleteProject(OpenShiftConfig.namespace(), false);
+        }
     }
 
     @Override
     public void testPlanExecutionFinished(TestPlan testPlan) {
         if (JUnitConfig.cleanOpenShift()) {
-            boolean deleted = openShift.deleteProject();
-            // For multi-module maven projects, other modules may attempt to crate project requests immediately after this modules deleteProject
-            if (deleted) {
-                BooleanSupplier bs = () -> openShift.getProject() == null;
-                new SimpleWaiter(bs, TimeUnit.MINUTES, 2, "Waiting for old project deletion").waitFor();
+            deleteProject(XTFConfig.get(OPENSHIFT_NAMESPACE), false);
+        }
+    }
+
+    private void deleteProject(String namespace, boolean waitForDeletion) {
+        OpenShift openShift = OpenShifts.master(namespace);
+        boolean deleted = openShift.deleteProject();
+        log.info("Start deleting namespace: " + openShift.getNamespace());
+        if (!deleted && waitForDeletion) {
+            BooleanSupplier bs = () -> openShift.getProject() == null;
+            new SimpleWaiter(bs, TimeUnit.MINUTES, 2, "Waiting for " + openShift.getNamespace() + " project deletion")
+                    .waitFor();
+        }
+    }
+
+    @Override
+    public FilterResult apply(TestDescriptor testDescriptor) {
+        if (testDescriptor instanceof MethodBasedTestDescriptor) {
+            boolean disabled = Arrays.stream(((MethodBasedTestDescriptor) testDescriptor).getTestClass().getAnnotations())
+                    .filter(annotation -> annotation instanceof Disabled).count() > 0;
+            if (!disabled) {
+                TestCaseContext.getTestcaseToNamespaceMap().putIfAbsent(
+                        ((MethodBasedTestDescriptor) testDescriptor).getTestClass().getName(),
+                        getNameSpaceForTestClass(testDescriptor));
+
+            }
+        }
+        return FilterResult.included(testDescriptor.getDisplayName());
+    }
+
+    private String getNameSpaceForTestClass(TestDescriptor testDescriptor) {
```

Comment:
This is related to other question regarding `TestCaseContext` responsibilities? Have you considered to add this method to `TestCaseContext.java`

Reply:
This one is specific for `ProjectCreator` class as it's used just in it when creating names (on namespaces) for test cases.

Reply:
I would like to have a single place from where namespace for running test case can be retrieved. Currently it is `OpenShiftConfig.namespace()`. If `getNameSpaceForTestClass(..)` is added to `TestCaseContext` then it would bypass `OpenShiftConfig.namespace()` logic.

However you're right that `TestCaseContext` responsibilities are not defined. I think it should server just for the purpose of holding test case name but not to be tied with namespace. I think that `Map<String, String> testcaseToNamespaceMap` should not be in `TestCaseContext` and moved to `OpenshiftConfig`. I'll make some changes and push new way for review.

Reply:
Can it be set as resolved?